File: lecture14/exercise14-A/solution.cpp
// Exercise 14-A: Cookies

#include <algorithm>
#include <deque>
#include <iostream>
#include <sstream>
#include <string>

using namespace std;

// Functions

bool readline(deque<int> &v) {
    string line;
    if (!getline(cin, line)) {
    	return false;
    }

    stringstream ss(line);
    int value;

    v.clear();
    while (ss >> value) {
    	v.push_back(value);
    }

    sort(v.rbegin(), v.rend());
    return true;
}

int feed_children(deque<int> &children, deque<int> &cookies) {
    int count = 0;

    while (!children.empty() && !cookies.empty()) {
    	auto child  = children.front();
    	auto cookie = cookies.front();

	children.pop_front();
    	if (child <= cookie) {
    	    cookies.pop_front();
    	    count++;
	}
    }

    return count;
}

// Main execution

int main(int argc, char *argv[]) {
    deque<int> children;
    deque<int> cookies;

    while (readline(children) && readline(cookies)) {
	cout << feed_children(children, cookies) << endl;
    }

    return 0;
}

File: lecture14/exercise14-A/solution.py
#!/usr/bin/env python3

# Exercise 14-A: Cookies

import sys

# Functions

def readline():
    return sorted(map(int, sys.stdin.readline().split()), reverse=True)

def feed_children(children, cookies):
    count = 0

    while cookies and children:
        child  = children.pop(0)
        cookie = cookies[0]

        if child <= cookie:
            cookies.pop(0)
            count += 1

    return count

# Main execution

def main():
    while (children := readline()) and (cookies := readline()):
        print(feed_children(children, cookies))

if __name__ == '__main__':
    main()
